### PR TITLE
Docker & docker-swarm role and sample of use in the some roles

### PR DESCRIPTION
**Pull request by vjrj**
This PR is a proposal of how to use docker-swarm (and ideally helm+k8s too in the future) with our current roles to allow a legacy VM deploy and a docker based one. This sample include support for: collectory, cas, species-list,  solrcloud/zookeeper, cassandra, gatus, biocache-hub, biocache-service, ala-bie, bie-index, namematching and sensitive-data-service.

To deploy a cluster use the `docker-swarm.yml` playbook and to deploy the collectory stack, `docker-common.yml` to deploy common services (`nginx` and `postfix` right now) and other services the `collectory-docker.yml` or `auth-docker.yml`, etc:

```
[docker_swarm]
# I add the variables here, because I have the dev group inventory in the same file
docker-cluster-2023-1 zookeeper_hosts=3 solr_hosts=3
docker-cluster-2023-2 zookeeper_hosts=3 solr_hosts=3
docker-cluster-2023-3 zookeeper_hosts=3 solr_hosts=3

[docker_swarm_managers]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[docker_swarm_workers]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[collectory]
# This deploy /data in a shared volume in the cluster using the collectory:vars of our inventories
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[cas-servers]
# Logs goes to /var/log/atlas and should be created and mounted
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[species-lists]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[zookeeper]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[solrcloud]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[cassandra3]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[biocache-service-clusterdb]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[gatus]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[biocache-hub]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[bie-index]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[bie-hub]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[namematching-service]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[sensitive-data-service]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

[portainer]
docker-cluster-2023-1
docker-cluster-2023-2
docker-cluster-2023-3

```

Or to test locally (if your desktop is a ubuntu or debian based distro as currently tries to install i18n package and do `apt-update` probably easy to skip):
```
[docker_swarm_dev]
127.0.0.1

[docker_swarm_managers_dev]
127.0.0.1

[docker_swarm_workers_dev]
127.0.0.1

[collectory]
127.0.0.1

[cas-servers]
127.0.0.1

[species-lists]
127.0.0.1

[solrcloud]
127.0.0.1

[zookeeper]
127.0.0.1

[cassandra3]
127.0.0.1

[biocache-service-clusterdb]
127.0.0.1

[gatus]
127.0.0.1

[biocache-hub]
127.0.0.1

[bie-hub]
127.0.0.1

[bie-index]
127.0.0.1

[namematching-service]
127.0.0.1

[sensitive-data-service]
127.0.0.1

[portainer]
127.0.0.1
``` 

Also:
```
[all:vars]
i18n_package_enabled=true # or false to use or not the i18n package
# This should create internal aliases to allow the internal communication using full urls
nginx_docker_internal_aliases=["colecciones-es.l-a.site", "auth-es.l-a.site", "listas-es.l-a.site", "status.l-a.site", "registros-es.l-a.site", "especies-es.l-a.site", "especies-ws-es.l-a.site", "namematching-service-es.l-a.site", "sensitive-data-service-es.l-a.site"]
# This mounts /data/letsencrypt in the nginx container
use_letsencrypt=true
# Optionally during development use this image with extra packages to debug networking, etc
docker_nginx_image=livingatlases/nginx-net-debug:latest
# To use internal dns of docker to resolve which upstream servers are up on each role
nginx_docker_swarm_resolver=true
# By default use /data/docker/{{ hostname }} but in other cases I tested with:
# docker_data_dir=/var/lib/docker
gatus_path=/
gatus_url=status.l-a.site
use_gatus=true
docker_default_user=ubuntu

[docker_swarm:vars]
mongodb_version=6.0
docker_solr_hosts = [ "docker-cluster-2023-1", "docker-cluster-2023-2", "docker-cluster-2023-3" ]
zookeeper_hosts=3
solr_hosts=3
docker_development_mode=false

[docker_swarm_dev:vars]
ansible_connection=local
docker_default_user=youruser
docker_development_mode=true
zookeeper_hosts=1
solr_hosts=1

```

I'm reusing the mysql/mongo creation/insert/dump-restore tasks via opening a tunnel for the internal ports (mysql, mongo, cas...). and later closing them. With this we can reuse the same code in the vm-tasks and in the docker-tasks. The previous approach to use mysql/mongo init scripts didn't work as they only run at first db initialisation. 

The generated configurations can be used with kompose.io to have a base helm charts and substitute the same variable names. 

Run with something like:
``` 
cd your-inventories; 
ansible-playbook -u ubuntu -i la-test-inventory.ini $ALA_INST/ansible/docker-swarm.yml $ALA_INST/ansible/docker-common.yml  $ALA_INST/ansible/auth-docker.yml $ALA_INST/ansible/collectory-docker.yml $ALA_INST/ansible/species-list-docker.yml  $ALA_INST/ansible/solrcloud-docker.yml ... 
``` 
or only the docker-services with the tag `--tags docker-service`  or run all currently supported services with:
```
ansible-playbook -u ubuntu -i la-test-inventory.ini $ALA_INST/ansible/docker-swarm-main.yml  
```

You can add `--extra-vars docker_stack_rm=false` to remove each stack before deploying it (useful during development).

Issues:
- [ ] cas-management redirects to localhost incorrectly (I need help here): https://auth-es.l-a.site/cas/login?service=https%3A%2F%2Flocalhost%3A8443%2Fcas-management%2Fcallback%3Fclient_name%3DCasClient
- [x] nginx default config does not include sites-enabled. Solution: a) to create a new custom nginx image that add it to the conf. b) update de nginx role to generate only the conf for the docker container and mount it. **Following a)**
- [x] Nginx does not boot if some service is down because it not resolve the direction in the proxy call:
```
common_nginx.1.pwgqie9eyi8l@gbif-es-docker-cluster-2023-2    | 2023/08/22 00:29:02 [emerg] 1#1: host not found in upstream "cas_management" in /etc/nginx/sites-enabled/auth-es.l-a.site.conf:76
common_nginx.1.pwgqie9eyi8l@gbif-es-docker-cluster-2023-2    | nginx: [emerg] host not found in upstream "cas_management" in /etc/nginx/sites-enabled/auth-es.l-a.site.conf:76
```
in other words, we need a dynamic config for upstreams that allows load balancing etc (like to use `resolver 127.0.0.11 ipv6=off;` or docker-gen or other solutions). **Using resolver for now.**
- [x] Key added in a deprecated way:
 W: https://download.docker.com/linux/ubuntu/dists/jammy/InRelease: Key is stored in legacy trusted.gpg keyring (/etc/apt/trusted.gpg), see the DEPRECATION section in apt-key(8) for details.
- [x] Configure scalability to allow, for instance, to run nginx in all the 80/443 ports of the cluster elements
- [ ] The postfix docker-compose configuration is not tested and probably needs more work
- [x] To add sorl_config to create bie collection
- [x] Force to run each solcloudr/zookeeper in a specific cluster host
- [x] Auto-configure a status service like https://github.com/TwiN/gatus (the configuration can be generated)
- [ ] Portainer support (almost ready)
- [ ] Create bie collections as with biocache
- [ ] userdetails_base_url in auth and alerts in without /userdetails ... and in the rest is with /userdetails
- [x] allow extra gatus checks (solr/cassandra and custom, via a dict variable per role and/or tasks)
- [ ] do some nginx reload, but docker based for updated vhosts

### CODE REVIEW

**Comment by vjrj** [comment]
Maybe, to distinguish from other software,  I should rename the services from:

- auth_cas > la_auth_cas
- auth_userdetails > la_auth_userdetails
- auth_mongo > la_auth_mongo
(...)
- collectory_mysql > la_collectory_mysql
- collectory_collectory > la_collectory_war ?
- common_nginx > la_common_nginx
- common_postfix > la_common_postfix (maybe we can use other world than “common”)

networks:
- auth_auth > la_auth_default
- la-internal > la_internal
- collectory_collectory > collectory_default

and maybe also something similar with volumes.

Currently:

image

Comments?

**Comment by vjrj** [comment]
If finally renamed the stacks and now looks better:

image

I also added the species-list service.

**Comment by vjrj** [comment]
Added solrcloud/zookeeper stack.

**Comment by vjrj** [comment]
Added cassandra and biocache-service docker stacks.

**Comment by vjrj** [comment]
I added biocache-hub and gatus monitoring support. The last one is quite interesting to check internally and externally the services status to fast detect issues even during development.

The configuration, if gatus is enabled is auto-generated with the last task of `nginx_vhost` and looks like this:

File: ansible/roles/collectory/vars/main.yml [hunk 1]
```diff
@@ -13,4 +13,6 @@ packaging: "war"
 
 collectory_war_url: "{{ collectory_maven_repo_ws_url | default(maven_repo_ws_url) }}"
 
-collectory_data_dir: "{{ data_dir }}/{{ collectory_app | default('generic-collectory') }}"
\ No newline at end of file
+collectory_data_dir: "{{ data_dir }}/{{ collectory_app | default('generic-collectory') }}"
```
Comment (djtfmartin) [inline ansible/roles/collectory/vars/main.yml:16]: generic-collectory is not something we support any more. Suggest just replacing with 'collectory'

**Review by djtfmartin — APPROVED** [review]
LGTM - suggest squash before merging

**Pull request merged**